# Retail Profitability Analysis - Data CLeaning

## 1. Load the CSV and Inspect the Data

* Load your Superstore dataset into Python using Pandas
* Check the dimensions, types, and quality of the data

In [1]:
# Libraries for data manipulation
import pandas as pd

# Libraries for exploratory data analysis and visualization
import matplotlib.pyplot as plt
import seaborn as sns

# Set visual style for consistency
sns.set(style="whitegrid")

In [2]:
# file path
file_path = '../data/Sample - Superstore.csv'

# Load the CSV with proper encoding to handle special characters
df = pd.read_csv(file_path, encoding="latin1")

# Quick look at the data
df.sample(10)

Row ID        Order ID  Order Date   Ship Date       Ship Mode  \
3789    3790  US-2016-133508   4/17/2016   4/21/2016  Standard Class   
7258    7259  CA-2016-152730   5/30/2016    6/4/2016  Standard Class   
3820    3821  CA-2017-141873   10/9/2017  10/14/2017  Standard Class   
9167    9168  CA-2016-100300   6/24/2016   6/26/2016    Second Class   
3998    3999  CA-2015-105627    3/8/2015   3/12/2015  Standard Class   
1803    1804  CA-2017-158379   9/22/2017   9/26/2017    Second Class   
9702    9703  CA-2017-147032   7/31/2017    8/4/2017  Standard Class   
2885    2886  US-2016-159856  10/18/2016  10/22/2016  Standard Class   
1978    1979  CA-2015-109939    5/8/2015   5/12/2015  Standard Class   
1753    1754  CA-2017-116715   12/2/2017   12/5/2017     First Class   

     Customer ID    Customer Name      Segment        Country           City  \
3789    SW-20350       Sean Wendt  Home Office  United States          Omaha   
7258    EM-14140     Eugene Moren  Home Office  United States       Superior   
3820    HG-14845     Harry Greene     Consumer  United States  New York City   
9167    MJ-17740        Max Jones     Consumer  United States      San Diego   
3998    DK-12895      Dana Kaydos     Consumer  United States        Kenosha   
1803    JA-15970     Joseph Airdo     Consumer  United States   Philadelphia   
9702    LB-16795   Laurel Beltran  Home Office  United States     Wilmington   
2885    EP-13915       Emily Phan     Consumer  United States          Tempe   
1978    AA-10375     Allen Armold     Consumer  United States          Salem   
1753    VW-21775  Victoria Wilson    Corporate  United States  San Francisco   

      ... Postal Code   Region       Product ID         Category Sub-Category  \
3789  ...       68104  Central  OFF-FA-10000134  Office Supplies    Fasteners   
7258  ...       54880  Central  OFF-AR-10000940  Office Supplies          Art   
3820  ...       10024     East  FUR-BO-10003034        Furniture    Bookcases   
9167  ...       92037     West  OFF-AP-10001242  Office Supplies   Appliances   
3998  ...       53142  Central  FUR-FU-10000308        Furniture  Furnishings   
1803  ...       19134     East  OFF-SU-10002881  Office Supplies     Supplies   
9702  ...       19805     East  OFF-PA-10003256  Office Supplies        Paper   
2885  ...       85281     West  FUR-CH-10003396        Furniture       Chairs   
1978  ...       97301     West  OFF-AR-10000127  Office Supplies          Art   
1753  ...       94122     West  OFF-PA-10004475  Office Supplies        Paper   

                                           Product Name     Sales  Quantity  \
3789                  Advantus Push Pins, Aluminum Head    29.050         5   
7258                                         Newell 343    14.700         5   
3820      O'Sullivan Elevations Bookcase, Cherry Finish   314.352         3   
9167   APC 7 Outlet Network SurgeArrest Surge Protector   241.440         3   
3998           Deflect-o Glass Clear Studded Chair Mats   373.080         6   
1803  Martin Yale Chadless Opener Electric Letter Op...  4663.736         7   
9702         Avery Personal Creations Heavyweight Cards    11.540         1   
2885                          Global Deluxe Steno Chair   307.920         5   
1978                                         Newell 321     5.248         2   
1753                                         Xerox 1940   109.920         2   

      Discount     Profit  
3789       0.0     9.0055  
7258       0.0     3.9690  
3820       0.2   -15.7176  
9167       0.0    72.4320  
3998       0.0    82.0776  
1803       0.2 -1049.3406  
9702       0.0     5.5392  
2885       0.2   -34.6410  
1978       0.2     0.5904  
1753       0.0    53.8608  

[10 rows x 21 columns]

In [3]:
# Number of rows and columns
print("Dataset shape:", df.shape)

Dataset shape: (9994, 21)


#### Observations
> The dataset contains 9994 rows and 21 columns.

In [4]:
# Check column data types and missing values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         9994 non-null   int64  
 1   Order ID       9994 non-null   object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9994 non-null   int64  
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quantity

In [5]:
# Count of missing/null values in each column
df.isnull().sum()

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64

#### Observations
> * Numeric columns (`Sales`, `Discount`, `Profit`, `Quantity`) are correctly loaded as floats/integers.
> * Categorical columns (`Category`,`Sub-Category`, `Segment`, `Region`) are strings
> * `Order Date` and `Ship Date`column needs to be converted to **datetime** for trend analysis.
> * There are **no missing values**, which simplifies preprocessing.

In [6]:
# Quick statistics for numeric columns
df.describe()

Row ID   Postal Code         Sales     Quantity     Discount  \
count  9994.000000   9994.000000   9994.000000  9994.000000  9994.000000   
mean   4997.500000  55190.379428    229.858001     3.789574     0.156203   
std    2885.163629  32063.693350    623.245101     2.225110     0.206452   
min       1.000000   1040.000000      0.444000     1.000000     0.000000   
25%    2499.250000  23223.000000     17.280000     2.000000     0.000000   
50%    4997.500000  56430.500000     54.490000     3.000000     0.200000   
75%    7495.750000  90008.000000    209.940000     5.000000     0.200000   
max    9994.000000  99301.000000  22638.480000    14.000000     0.800000   

            Profit  
count  9994.000000  
mean     28.656896  
std     234.260108  
min   -6599.978000  
25%       1.728750  
50%       8.666500  
75%      29.364000  
max    8399.976000

#### Observations
> * Sales and Profit have wide ranges:
>   * Sales: min 0.44, max 22,638
>   * Profit: min -6599.98, max 8,399.98
>   * This indicates some orders are loss-making and some are extremely high-value.
> * Discount ranges from 0 to 0.8, which is within expectations.
> * Quantity ranges from 1 to 14, consistent with typical order sizes.
> * There are outliers in Profit (negative values)that should be considered during analysis but not removed without business context.

In [7]:
# Summary statistics for categorical (object) columns
df.describe(include="object")

Order ID Order Date   Ship Date       Ship Mode Customer ID  \
count             9994       9994        9994            9994        9994   
unique            5009       1237        1334               4         793   
top     CA-2017-100111   9/5/2016  12/16/2015  Standard Class    WB-21850   
freq                14         38          35            5968          37   

        Customer Name   Segment        Country           City       State  \
count            9994      9994           9994           9994        9994   
unique            793         3              1            531          49   
top     William Brown  Consumer  United States  New York City  California   
freq               37      5191           9994            915        2001   

       Region       Product ID         Category Sub-Category     Product Name  
count    9994             9994             9994         9994             9994  
unique      4             1862                3           17             1850  
top      West  OFF-PA-10001970  Office Supplies      Binders  Staple envelope  
freq     3203               19             6026         1523               48

#### Observations
> * There are **5009 unique Order IDs across 9994 rows**, indicating that **each row represents a line item, not a complete order**. A single order can include multiple products, each recorded as a separate row. For order-level and customer-level analysis, metrics must be calculated accordingly.
> * There are **793 unique customers**, which is sufficient to analyze repeat purchases and customer profitability.
> * The dataset includes **3 customer segments: Consumer, Corporate, and Home Office.**
> * The Consumer segment dominates the dataset (5191 records), which may influence overall profitability trends.
> * There are 4 ship modes . Shipping mode may later be explored as a cost or operational factor, but is not a primary focus of this analysis.
> * Data covers 49 states, 531 cities, and 4 regions (West, East, Central, South).West is the most frequent region, suggesting potential regional imbalance in sales and profit contributions.
> * There are 3 product categories and 17 sub-categories.
> * All records belong to a single country (United States), eliminating the need for country-level filtering or normalization.
> * **`Order Date` and `Ship Date` are currently stored as strings**. These will need to be converted to datetime format for time-based analysis (trends, cohorts, yearly performance).

In [8]:
# check for duplicates
df.duplicated().sum()

np.int64(0)

> There are no duplicates.

## 2. Data Cleaning & Data Preprocessing

The goal of this step is :

* Correct datetime columns
* Standardized column names
* Drop unnecessary columnns
* One key derived metric: profit margin

In [9]:
# Standardize column names: lowercase, underscores, no spaces
df.columns = (
    df.columns
    .str.strip()
    .str.lower()
    .str.replace(' ', '_')
    .str.replace('-', '_')
)

df.columns

Index(['row_id', 'order_id', 'order_date', 'ship_date', 'ship_mode',
       'customer_id', 'customer_name', 'segment', 'country', 'city', 'state',
       'postal_code', 'region', 'product_id', 'category', 'sub_category',
       'product_name', 'sales', 'quantity', 'discount', 'profit'],
      dtype='object')

#### Observations
> * We standardized column names to remove spaces and mixed casing

In [10]:
# Convert order_date and ship_date to datetime
df['order_date'] = pd.to_datetime(df['order_date'])
df['ship_date'] = pd.to_datetime(df['ship_date'])

df[['order_date', 'ship_date']].dtypes

order_date    datetime64[ns]
ship_date     datetime64[ns]
dtype: object

#### Observations
> Order Date and Ship Date were converted from string to datetime format.  
This enables time-based analysis such as yearly trends, monthly performance, and cohort analysis.

In [11]:
# drop unnecessay columns
df = df.drop(columns=["row_id", "postal_code"])

In [12]:
# create new column
df["profit_margin"] = df["profit"] / df["sales"]

In [13]:
df['profit_margin'].describe()

count    9994.000000
mean        0.120314
std         0.466754
min        -2.750000
25%         0.075000
50%         0.270000
75%         0.362500
max         0.500000
Name: profit_margin, dtype: float64

#### Observarions
> * The average profit margin is ~0.12, while the median is 0.27 . This indicates a right-skewed distribution, where many orders are profitable,but a smaller number of heavily loss-making orders pull the average down.
> * Negative profit margins exist (minimum: -2.75). These occur when high discounts or costs exceed sales revenue.
> * Maximum profit margin is capped at 0.50.

## Save the file

In [14]:
df.to_csv("../data/superstore_processed.csv", index=False)